### PR TITLE
[BrowserKit] add a way to switch to ajax for one request

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | #20306 
| License       | MIT
| Doc PR        | will do

Follow the work on #20306.

/cc @fabpot is it the right implementation ?

### CODE REVIEW

File: src/Symfony/Component/BrowserKit/Client.php [hunk 1]
```diff
@@ -150,6 +150,16 @@ public function getServerParameter($key, $default = '')
         return isset($this->server[$key]) ? $this->server[$key] : $default;
     }
 
+    public function switchToAjax()
```
Comment (iltar) [inline src/Symfony/Component/BrowserKit/Client.php]: Wouldn't it make more sense to use the term `XmlHttpRequest` instead of `Ajax`?

Comment (Simperfit) [reply src/Symfony/Component/BrowserKit/Client.php]: I guess it would make more sense

File: src/Symfony/Component/BrowserKit/Client.php [hunk 2]
```diff
@@ -354,7 +374,7 @@ protected function doRequestInProcess($request)
             unlink($deprecationsFile);
             foreach ($deprecations ? unserialize($deprecations) : array() as $deprecation) {
                 if ($deprecation[0]) {
-                    trigger_error($deprecation[1], E_USER_DEPRECATED);
+                    @trigger_error($deprecation[1], E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/BrowserKit/Client.php]: this change is adding a bug, fabbot is not always right

File: src/Symfony/Component/BrowserKit/Client.php [hunk 3]
```diff
@@ -249,14 +259,15 @@ public function click(Link $link)
      *
      * @param Form  $form   A Form instance
      * @param array $values An array of form field values
+     * @param bool  $isAjax Whether it's ajax or not
      *
      * @return Crawler
      */
-    public function submit(Form $form, array $values = array())
+    public function submit(Form $form, array $values = array(), $isAjax = false)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/BrowserKit/Client.php]: adding arguments to a public method is a BC break, see https://symfony.com/bc

File: src/Symfony/Component/BrowserKit/Client.php (hunk 2 above)
Comment (Simperfit) [reply src/Symfony/Component/BrowserKit/Client.php]: will revert

File: src/Symfony/Component/BrowserKit/Client.php (hunk 3 above)
Comment (Simperfit) [reply src/Symfony/Component/BrowserKit/Client.php]: I understood. I have changed the implementation.

File: src/Symfony/Component/BrowserKit/Client.php [hunk 4]
```diff
@@ -150,6 +150,16 @@ public function getServerParameter($key, $default = '')
         return isset($this->server[$key]) ? $this->server[$key] : $default;
     }
 
+    public function switchToXMLHttpRequest()
+    {
+        $this->setServerParameter('HTTP_X-Requested-With', 'XMLHttpRequest');
```
Comment (dmaicher) [inline src/Symfony/Component/BrowserKit/Client.php]: shouldn't it be `'HTTP_X_REQUESTED_WITH'`?

Comment (Simperfit) [reply src/Symfony/Component/BrowserKit/Client.php]: The header name is really X-Requested-With, so no I guess.

**Comment by Simperfit** [comment]
fabpot failure seems unrelated

File: src/Symfony/Component/BrowserKit/Client.php (hunk 4 above)
Comment (dunglas) [reply src/Symfony/Component/BrowserKit/Client.php]: I think @dmaicher is right: PHP normalizes dash to low dashes when hydrating the $_SERVER parameter.

Comment (Simperfit) [reply src/Symfony/Component/BrowserKit/Client.php]: fixed, i've tested it and @dmaicher is actually right, sorry I did just check the header name.

**Comment by Simperfit** [comment]
This one is ready for review.

File: src/Symfony/Component/BrowserKit/Tests/ClientTest.php [hunk 5]
```diff
@@ -94,6 +94,16 @@ public function testGetRequest()
...
     }
 
+    public function testGetRequestWithXMLHttpRequest()
+    {
+        $client = new TestClient();
+        $client->switchToXMLHttpRequest();
+        $client->request('GET', 'http://example.com/', array(), array(), array(), null, true, true);
+        $this->assertEquals($client->getRequest()->getServer()['HTTP_X_REQUESTED_WITH'], 'XMLHttpRequest');
+        $client->removeXMLHttpRequest();
+        $this->assertEquals('http://example.com/', $client->getRequest()->getUri(), '->getCrawler() returns the Request of the last request');
```
Comment (dmaicher) [inline src/Symfony/Component/BrowserKit/Tests/ClientTest.php]: This should make sure the server param is not set anymore instead?

File: src/Symfony/Component/BrowserKit/Client.php [hunk 6]
```diff
@@ -150,6 +150,16 @@ public function getServerParameter($key, $default = '')
         return isset($this->server[$key]) ? $this->server[$key] : $default;
     }
 
+    public function switchToXMLHttpRequest()
+    {
+        $this->setServerParameter('HTTP_X_REQUESTED_WITH', 'XMLHttpRequest');
+    }
+
+    public function removeXMLHttpRequest()
```
Comment (dmaicher) [inline src/Symfony/Component/BrowserKit/Client.php]: I find this name a bit confusing. We are not removing any request :)

Maybe we can just only have a setter instead? `setXMLHttpRequest($isXMLHttpRequest)`

Comment (Simperfit) [reply src/Symfony/Component/BrowserKit/Client.php]: I will add Header a the end I think

**Comment by Simperfit** [comment]
Named changed and test added.

fabbot error is wrong.

**Comment by Simperfit** [comment]
AppVeyor failure is unrelated and fabbot error is a false positive.

@fabpot WDYT of this new implem ?

**Comment by Simperfit** [comment]
PR rebased with master

File: src/Symfony/Component/BrowserKit/Tests/ClientTest.php [hunk 7]
```diff
@@ -94,6 +94,17 @@ public function testGetRequest()
...
 
+    public function testGetRequestWithXMLHttpRequest()
+    {
+        $client = new TestClient();
+        $client->switchToXMLHttpRequest();
+        $client->request('GET', 'http://example.com/', array(), array(), array(), null, true, true);
+        $this->assertEquals($client->getRequest()->getServer()['HTTP_X_REQUESTED_WITH'], 'XMLHttpRequest');
+        $client->removeXMLHttpRequestHeader();
+        $this->assertFalse($client->getServerParameter('HTTP_X_REQUESTED_WITH', false));
+        $this->assertEquals('http://example.com/', $client->getRequest()->getUri(), '->getCrawler() returns the Request of the last request');
```
Comment (stof) [inline src/Symfony/Component/BrowserKit/Tests/ClientTest.php]: this last assertion looks useless

Comment (Simperfit) [reply src/Symfony/Component/BrowserKit/Tests/ClientTest.php]: that's right, i've removed it.

File: src/Symfony/Component/BrowserKit/Client.php [hunk 8]
```diff
@@ -150,6 +150,16 @@ public function getServerParameter($key, $default = '')
         return isset($this->server[$key]) ? $this->server[$key] : $default;
     }
 
+    public function switchToXMLHttpRequest()
```
Comment (fabpot) [inline src/Symfony/Component/BrowserKit/Client.php]: What about `switchToXHR` instead?

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/BrowserKit/Client.php (hunk 8 above)
Comment (Simperfit) [reply src/Symfony/Component/BrowserKit/Client.php]: done

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @Simperfit.

**Comment by javiereguiluz** [comment]
I know ... it's too late to talk about this because it's merged ... but I wanted to say that I don't like the current implementation :(  It modifies a global estate with `switchToXHR()` and you must remember later to reset the original estate with `removeXHR()`.

As an alternative, I wonder if we could introduce a new `xhrRequest()` method that wraps the `request()` method and it works like it, but does the XHR stuff internally just for that request. Thanks!

**Comment by Simperfit** [comment]
@javiereguiluz I agree with this statement, however I think we should let both switchToXHR and removeXHR public, it can be useful to be able to do it by ourself.

@fabpot WDYT ?

**Comment by fabpot** [comment]
I totally agree with @javiereguiluz here. I was too fast to merge. The description of the PR si misleading.

@Simperfit Can you submit a PR to implement what's on the tin?

**Comment by Simperfit** [comment]
Yes I will be doing it today !
Le ven. 2 mars 2018 à 14:14, Fabien Potencier <notifications@github.com> a
écrit :

**Pull request merged**